odbc_fdw.c

```diff
@@ -155,9 +182,208 @@ odbc_fdw_handler(PG_FUNCTION_ARGS)
+ */
+static char*
+empty_string_if_null(char *string)
+{
+	static char* empty_string = "";
+	return string == NULL ? empty_string : string;
+}
+
+static void
+extract_odbcFdwOptions(List *options_list, odbcFdwOptions *extracted_options)
+{
+	ListCell        *lc;
+	List            *mapping_list;
+
+	#ifdef DEBUG
+		elog(DEBUG1, "extract_init_odbcFdwOptions");
+	#endif
+
+	init_odbcFdwOptions(extracted_options);
+
+	/* Loop through the options, and get the foreign table options */
+	foreach(lc, options_list)
+	{
+		DefElem *def = (DefElem *) lfirst(lc);
+
+		if (strcmp(def->defname, "dsn") == 0)
+		{
+			extracted_options->dsn = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "driver") == 0)
+		{
+			extracted_options->driver = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "host") == 0)
+		{
+			extracted_options->host = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "port") == 0)
+		{
+			extracted_options->port = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "database") == 0)
+		{
+			extracted_options->database = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "schema") == 0)
+		{
+			extracted_options->schema = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "table") == 0)
+		{
+			extracted_options->table = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "prefix") == 0)
+		{
+			extracted_options->prefix = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "sql_query") == 0)
+		{
+			extracted_options->sql_query = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "sql_count") == 0)
+		{
+			extracted_options->sql_count = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "username") == 0)
+		{
+			extracted_options->username = defGetString(def);
+			continue;
+		}
+
+		if (strcmp(def->defname, "password") == 0)
+		{
+			extracted_options->password = defGetString(def);
+			continue;
+		}
+
+		/* Column mapping goes here */
+		/* TODO: is this useful? if so, how can columns names coincident
+		   with option names be escaped? */
+		extracted_options->mapping_list = lappend(extracted_options->mapping_list, def);
+	}
+
+	/* Convert empty strings to NULL pointers */
+	normalize_empty_string(&(extracted_options->dsn));
+	normalize_empty_string(&(extracted_options->driver));
+	normalize_empty_string(&(extracted_options->host));
+	normalize_empty_string(&(extracted_options->port));
+	normalize_empty_string(&(extracted_options->database));
+	normalize_empty_string(&(extracted_options->schema));
+	normalize_empty_string(&(extracted_options->table));
+	normalize_empty_string(&(extracted_options->prefix));
+	normalize_empty_string(&(extracted_options->sql_query));
+	normalize_empty_string(&(extracted_options->sql_count));
+	normalize_empty_string(&(extracted_options->username));
+	normalize_empty_string(&(extracted_options->password));
```

Comment:
I don't like it but just got it by seeing it in full context: the point is being able to manage optional params and empty params.

I'd rather prefer a function like:
```c
inline int str_empty_or_null(char *s) {
    if (s == NULL || s[0] == '\0')
        return 1;
    else
        return 0;
}
```

And I'll make my case: why? because otherwise the processing of the strings is pretty disconnected of the reasons to do so. Makes the code harder to understand and does not reflect that the normalize function is to be called at a certain point, in a certain order: initialize, gather params, normalize, check paramas.